core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -142,10 +142,18 @@ protected WritePrimaryResult shardOperationOnPrimary(BulkShardRequest request, I
                 case INDEX:
                     final IndexRequest indexRequest = (IndexRequest) itemRequest;
                     Engine.IndexResult indexResult = executeIndexRequestOnPrimary(indexRequest, primary, mappingUpdatedAction);
-                    operationResult = indexResult;
-                    response = indexResult.hasFailure() ? null
-                            : new IndexResponse(primary.shardId(), indexRequest.type(), indexRequest.id(),
+                    if (indexResult.hasFailure() == false) {
+                        // update the version on request so it will happen on the replicas
+                        final long version = indexResult.getVersion();
+                        indexRequest.version(version);
+                        indexRequest.versionType(indexRequest.versionType().versionTypeForReplicationAndRecovery());
+                        assert indexRequest.versionType().validateVersionForWrites(indexRequest.version());
+                        response = new IndexResponse(primary.shardId(), indexRequest.type(), indexRequest.id(),
                                 indexResult.getVersion(), indexResult.isCreated());
+                    } else {
+                        response = null;
+                    }
+                    operationResult = indexResult;
                     replicaRequest = request.items()[requestIndex];
```

Comment:
`replicaRequest` is a `BulkItemRequest` so we have to set it to `request.item()[requestIndex]` instead of `indexRequest`